Hide all symbols (except stable/headeronly/shim) if TORCH_STABLE_ONLY is defined

mikaylagawarecki opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/161660

This extends the `TORCH_STABLE_ONLY` stopgap added in https://github.com/pytorch/pytorch/pull/161658

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #169307
* __->__ #167496

mikaylagawarecki committed:
Hide all symbols (except stable/headeronly/shim) if TORCH_STABLE_ONLY is defined

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Hide all symbols (except stable/headeronly/shim) if TORCH_STABLE_ONLY is defined"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Hide all symbols (except stable/headeronly/shim) if TORCH_STABLE_ONLY is defined"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Hide all symbols (except stable/headeronly/shim) if TORCH_STABLE_ONLY is defined"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Hide all symbols (except stable/headeronly/shim) if TORCH_STABLE_ONLY is defined"

Fixes https://github.com/pytorch/pytorch/issues/161660

This extends the `TORCH_STABLE_ONLY` stopgap added in https://github.com/pytorch/pytorch/pull/161658




[ghstack-poisoned]

janeyx99 reviewed (commented):
Approach makes sense to me. I have two higher level thoughts:
1. We want folks to only have to define one compiler flag for stable ABI. Currently, that's defining TORCH_TARGET_VERSION, so could we make it so that either TORCH_STABLE_ONLY is turned on if TORCH_TARGET_VERSION is set? So people don't have to define both?

2. libtorch_agnostic_extension setup.py should be changed to use whatever single UX we decide to go with!

mikaylagawarecki committed:
Update on "Hide all symbols (except stable/headeronly/shim) if TORCH_STABLE_ONLY is defined"

Fixes https://github.com/pytorch/pytorch/issues/161660

This extends the `TORCH_STABLE_ONLY` stopgap added in https://github.com/pytorch/pytorch/pull/161658




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Hide all symbols (except stable/headeronly/shim) if TORCH_STABLE_ONLY is defined"

Fixes https://github.com/pytorch/pytorch/issues/161660

This extends the `TORCH_STABLE_ONLY` stopgap added in https://github.com/pytorch/pytorch/pull/161658




[ghstack-poisoned]

janeyx99 commented on test/cpp_extensions/libtorch_agnostic_extension/setup.py:
```diff
@@ -33,7 +33,7 @@ def run(self):
 def get_extension():
     extra_compile_args = {
-        "cxx": ["-fdiagnostics-color=always"],
+        "cxx": ["-fdiagnostics-color=always", "-DTORCH_STABLE_ONLY"],
```
I am missing something -- shouldn't we hide symbols for any wheel we build of torch? And doesn't python setup.py develop build a wheel anyway now?

mikaylagawarecki replied:
@janeyx99 Ok seems like it's just a local thing, but CI will fail as expected https://hud.pytorch.org/pr/pytorch/pytorch/167594#55137402027

mikaylagawarecki committed:
Update on "Hide all symbols (except stable/headeronly/shim) if TORCH_STABLE_ONLY is defined"

Fixes https://github.com/pytorch/pytorch/issues/161660

This extends the `TORCH_STABLE_ONLY` stopgap added in https://github.com/pytorch/pytorch/pull/161658




[ghstack-poisoned]

janeyx99 commented on setup.py:
```diff
@@ -1358,6 +1358,45 @@ def __exit__(self, *exc_info: object) -> None:
 # Need to create the proper LICENSE.txt for the wheel
 class bdist_wheel(setuptools.command.bdist_wheel.bdist_wheel):
+    def _wrap_headers_with_macro(self, bdist_dir: Path) -> None:
+        """Wrap all header files with #ifndef TORCH_STABLE_ONLY macro.
```
update doc

janeyx99 reviewed (approved):
LGTM, we should make sure somehow that TORCH_TARGET_VERSION  works too to hide the symbols.

Also cc @malfet for whether setup.py bdist_wheel is the best place to put or whether there's a broader solution.

mikaylagawarecki committed:
Update on "Hide all symbols (except stable/headeronly/shim) if TORCH_STABLE_ONLY is defined"

Fixes https://github.com/pytorch/pytorch/issues/161660

This extends the `TORCH_STABLE_ONLY` stopgap added in https://github.com/pytorch/pytorch/pull/161658




[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge

mikaylagawarecki committed:
Update on "Hide all symbols (except stable/headeronly/shim) if TORCH_STABLE_ONLY is defined"

Fixes https://github.com/pytorch/pytorch/issues/161660

This extends the `TORCH_STABLE_ONLY` stopgap added in https://github.com/pytorch/pytorch/pull/161658




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Hide all symbols (except stable/headeronly/shim) if TORCH_STABLE_ONLY is defined"

Fixes https://github.com/pytorch/pytorch/issues/161660

This extends the `TORCH_STABLE_ONLY` stopgap added in https://github.com/pytorch/pytorch/pull/161658




[ghstack-poisoned]

mikaylagawarecki commented:
@jeanschmidt is this ok to reland as long as CI is green?

mikaylagawarecki committed:
Update on "Hide all symbols (except stable/headeronly/shim) if TORCH_STABLE_ONLY is defined"

Fixes https://github.com/pytorch/pytorch/issues/161660

This extends the `TORCH_STABLE_ONLY` stopgap added in https://github.com/pytorch/pytorch/pull/161658




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Hide all symbols (except stable/headeronly/shim) if TORCH_STABLE_ONLY is defined"

Fixes https://github.com/pytorch/pytorch/issues/161660

This extends the `TORCH_STABLE_ONLY` stopgap added in https://github.com/pytorch/pytorch/pull/161658




[ghstack-poisoned]

mikaylagawarecki committed:
Update

[ghstack-poisoned]

atalman reviewed (approved):
lgtm

mikaylagawarecki committed:
Update on "Hide all symbols (except stable/headeronly/shim) if TORCH_STABLE_ONLY is defined"

Fixes https://github.com/pytorch/pytorch/issues/161660

This extends the `TORCH_STABLE_ONLY` stopgap added in https://github.com/pytorch/pytorch/pull/161658




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Hide all symbols (except stable/headeronly/shim) if TORCH_STABLE_ONLY is defined"

Fixes https://github.com/pytorch/pytorch/issues/161660

This extends the `TORCH_STABLE_ONLY` stopgap added in https://github.com/pytorch/pytorch/pull/161658




[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge -f "CI was already green before factoring out tests into above PR, xpu triton failures are unrelated and rocm wheel has been queueing for 10+ hours"
